### ISSUE TITLE
postgresql init script parsing breaks with functions

### ISSUE THREAD

**Issue by jeacott1** [testcontainers/testcontainers-java#608]
I'm trying to use the testcontainers postgres module with the TC_INITSCRIPT option, but am running into this issue

https://groups.google.com/forum/#!topic/mybatis-user/2WFKfpHub30

for statements like (taken from the ibatis example above)
```
CREATE FUNCTION update_family_timestamp() RETURNS trigger AS $$ 
BEGIN 
    NEW.time_updated := current_timestamp; 
    RETURN new; 
END; 
$$ LANGUAGE plpgsql; 
```

manifests an exception like:
```
	at org.testcontainers.jdbc.ext.ScriptUtils.executeSqlScript(ScriptUtils.java:309) ~[jdbc-1.6.0.jar:na]
	at org.testcontainers.jdbc.ext.ScriptUtils.executeSqlScript(ScriptUtils.java:216) ~[jdbc-1.6.0.jar:na]
	at org.testcontainers.jdbc.ContainerDatabaseDriver.runInitScriptIfRequired(ContainerDatabaseDriver.java:233) ~[jdbc-1.6.0.jar:na]
	at org.testcontainers.jdbc.ContainerDatabaseDriver.connect(ContainerDatabaseDriver.java:156) ~[jdbc-1.6.0.jar:na]
	at org.apache.tomcat.jdbc.pool.PooledConnection.connectUsingDriver(PooledConnection.java:310) ~[tomcat-jdbc-8.5.23.jar:na]
	... 98 common frames omitted
Caused by: java.lang.ArrayIndexOutOfBoundsException: null
	at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:597) ~[na:1.8.0_131]
	at java.lang.StringBuilder.append(StringBuilder.java:190) ~[na:1.8.0_131]
	at org.postgresql.core.Parser.parseSql(Parser.java:1020) ~[postgresql-9.4.1212.jre7.jar:9.4.1212.jre7]
	at org.postgresql.core.Parser.replaceProcessing(Parser.java:972) ~[postgresql-9.4.1212.jre7.jar:9.4.1212.jre7]
	at org.postgresql.core.CachedQueryCreateAction.create(CachedQueryCreateAction.java:40) ~[postgresql-9.4.1212.jre7.jar:9.4.1212.jre7]
	at org.postgresql.core.QueryExecutorBase.createQueryByKey(QueryExecutorBase.java:309) ~[postgresql-9.4.1212.jre7.jar:9.4.1212.jre7]
	at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:281) ~[postgresql-9.4.1212.jre7.jar:9.4.1212.jre7]
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:266) ~[postgresql-9.4.1212.jre7.jar:9.4.1212.jre7]
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:262) ~[postgresql-9.4.1212.jre7.jar:9.4.1212.jre7]
	at org.testcontainers.jdbc.ext.ScriptUtils.executeSqlScript(ScriptUtils.java:270) ~[jdbc-1.6.0.jar:na]
	... 102 common frames omitted
```

There needs to be a sendFullScript or similar option.

this code works and could be a simple drop in replacement https://github.com/mybatis/mybatis-3/blob/master/src/main/java/org/apache/ibatis/jdbc/ScriptRunner.java

**Comment by kiview** [comment]
Hey @jeacott1, if you already know a working solution, would you like to provide a PR? :slightly_smiling_face:

**Comment by jeacott1** [comment]
if you're happy for me to ditch the existing parser sure.
although gradle + eclipse is horrible, makes working on this thing terrible.

**Comment by rnorth** [comment]
Actually I’m about to merge a PR that fixes this! I’ll link the ticket later. 

Richard

**Comment by rnorth** [comment]
See #570 and #579 - if you have time, it would be fantastic if you could have a look at the PR and/or try it out.

Sorry if working with Eclipse is a PITA - I'd be keen to know if there's anything we could change to make things better (apart from going back to Maven!)

**Comment by jeacott** [comment]
@rnorth actually I dont think thats enough just looking at it.
I dont think it could handle arbitrary $$ delimiters or complex functions?, eg

```
CREATE OR REPLACE FUNCTION getitemname(itemid bigint) RETURNS text
    LANGUAGE plpgsql STRICT
    AS $$
DECLARE
    item RECORD;
BEGIN  

    WITH RECURSIVE itemtree (item_id, item_name, item_code,item_parentid,depth) AS (
            SELECT item_id,item_name, item_code,item_parentid,1 FROM om_item WHERE item_id = $1 AND deleted = 0
            UNION
            SELECT child.item_id,child.item_name, child.item_code,child.item_parentid,depth+1 FROM om_item child
            INNER JOIN itemtree parent ON child.item_id = parent.item_parentid 
            )

    SELECT array_to_string(array_agg(T.item_name), '>>>') INTO item FROM (SELECT * FROM itemtree ORDER BY depth DESC) T;
    return item;

END;
$$;
```

I think is by far safest and easiest to just pass the whole block to postgres and let it deal with it.

**Comment by rnorth** [comment]
I think that's worth exploring - AFAICT the only (?) way to pass the entire script to the DB is via the `/docker-entrypoint-initdb.d` folder, so it'll be a bit of work. However, probably worthwhile for all the databases we can support this way.

For now I'm inclined to go ahead with #579 as a 'better than nothing' improvement. It can't handle the `$$` delimiter but it's a step better.

Let's leave this issue open to address separately by redoing the passthrough mechanism.

**Comment by jeacott** [comment]
@rnorth i have a working version, and ill make  pr available when i get a chance. its not that complex at all actually, but i dont want to impact other decisions youve made or might have on the roadmap, so take what i provide as guidance of a workable solution and do with it as you like.

**Comment by jeacott1** [comment]
see a PR here for an approach https://github.com/testcontainers/testcontainers-java/pull/624

**Comment by jeacott** [comment]
see a pr here for an approach #624

**Comment by rnorth** [comment]
#579 was released in [1.10.4](https://github.com/testcontainers/testcontainers-java/releases/tag/1.10.4) as a basic approach.

**Comment by cvgaviao** [comment]
@rnorth, while I was trying to find an answer to my [issue](https://github.com/testcontainers/testcontainers-java/issues/2814) I found this one. maybe there is a regression?
